src/execution.rs

```diff
@@ -291,11 +330,63 @@ impl<'a> ScopedVariables<'a> {
     }
 }
 
+/// Excerpts of source from either the target language file or the tsg rules file.
+pub struct Excerpt<'a> {
+    path: &'a Path,
+    source: Option<&'a str>,
+    location: Location,
+}
+
+impl<'a> Excerpt<'a> {
+    pub fn from_source(path: &'a Path, source: &'a str, location: Location) -> Excerpt<'a> {
+        Excerpt {
+            path,
+            source: source.lines().nth(location.row),
+            location: location,
+        }
+    }
+
+    fn gutter_width(&self) -> usize {
+        ((self.location.row + 1) as f64).log10() as usize + 1
+    }
+}
+
+impl<'a> std::fmt::Display for Excerpt<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{}:{}:\n{}{}{}\n{}{}{}{}",
+            // path and line/col
+            Colour::White
+                .bold()
+                .paint(self.path.to_str().unwrap_or("<unknown file>")),
+            Colour::White
+                .bold()
+                .paint(format!("{}", self.location.row + 1)),
+            Colour::White
+                .bold()
+                .paint(format!("{}", self.location.column + 1)),
+            // first line: line number & source
+            Colour::Blue.paint(format!("{}", self.location.row + 1)),
+            Colour::Blue.paint(" | "),
+            self.source.unwrap_or("<no source found>"),
+            // second line: caret
+            " ".repeat(self.gutter_width()),
+            Colour::Blue.paint(" | "),
+            " ".repeat(self.location.column),
+            Colour::Green.bold().paint("^")
+        )
+    }
+}
+
 impl Stanza {
-    fn execute<'g, 'l, 's, 'tree>(
+    fn execute<'a, 'g, 'l, 's, 'tree>(
         &self,
         tree: &'tree Tree,
+        source_path: &'tree Path,
         source: &'tree str,
+        tsg_path: &'a Path,
+        tsg_source: &'a str,
```

Comment:
In a past life (one commit ago at time of writing), these were all added to the `ExecutionContext` along with `source`. Happily, since we're contextualizing the errors in this very function, we don't need to pass them around to every `evaluate` definition.

(`source` still has to be carried around in order to evaluate calls.)